---
title: 'Cell Type Distribution Score: A method to measure the overall cell type composition
  of samples at a single cell level'
author: "Tanya Karagiannis, Stefano Monti, Paola Sebastiani"
output:
  html_document:
    theme: united
    code_folding: show
    toc: true
    toc_float: true
    toc_collapse: false
---

# Introduction

The Cell Type Distribution Score is an entropy-based score to measure and assess the overall cell type composition across samples from single cell transcriptomic datasets.

Install the CTDS.score function by cloning the github repository and load the function in R.

```{r}
source('~/CTDS/R/distribution_score.R')

```


# Usage

We demonstrate the implementation of this method on cell type composition level data from three single cell RNA-seq datasets of aging and longevity. In this example, we apply the cell type distribution score on the normalized cell type proportions per sample and compare differences in distribution score between four age groups across the human lifespan to assess overall cell type composition in aging and extreme longevity.



Load libraries
```{r, include=TRUE, messages=FALSE, warnings=FALSE}
suppressMessages(c(library(dplyr),
                   library(ggplot2),
                   library(Seurat),
                   library(SingleCellExperiment),
                   library(reshape2)))
```

Load integrated single cell RNA-seq datasets of PBMCs from aging and extreme longevity individuals

```{r}
pbmc.combined <- readRDS(file = "~/Data/pbmc.combined.rds")

```

## Creating matrix of normalized cell type proportions

Create table of normalized cell type proportions per sample (samples v. cell types) from Seurat object.
```{r}
ct_prop <- prop.table(table(pbmc.combined@meta.data$sample.ID, pbmc.combined@meta.data$ct.consensus), margin = 1)

head(ct_prop)
# require(knitr)
# require(kableExtra)
# kable(ct_prop, caption = paste0("cell type proportions per sample"),"html") %>% kable_styling("striped") %>% scroll_box(width = "100%",height = "300px")
```

Create matrix of metadata information from Seurat object.
```{r}

metadat <- lapply(1:nrow(ct_prop),function(x){
  metaInfo <- unique(pbmc.combined@meta.data[pbmc.combined@meta.data[,"sample.ID"] %in% rownames(ct_prop)[x],c("age.quartiles", "sex")])
  rownames(metaInfo) <- rownames(ct_prop)[x]
  return(metaInfo)
  })

metadat <- do.call(rbind, metadat)

head(metadat)
```
  
  
Visualize cell type proportions for all samples grouped by age group.
```{r, fig.show="hold", out.width="50%"}
normprop <- cbind.data.frame(ct_prop, metadat)
colnames(normprop)[1:2] <- c("sample.ID","cell.type")

normprop$cell.type <- factor(normprop$cell.type, levels=c("CD4 T NonCytotoxic","CD4 T Cytotoxic","CD8 T","T gamma delta", "NK","BC Memory","BC Naive","M14","M16","mDC","pDC","Plasma"))

normprop$age.quartiles <- factor(normprop$age.quartiles, levels=c("20-39","40-59","60-89","Extreme Longevity"))

age.groups <- levels(normprop$age.quartiles)


for (x in 1:length(age.groups)){
  dat <- normprop %>% dplyr::filter(age.quartiles == age.groups[x])

  p4 <- ggplot(dat, aes(interaction(sample.ID, age.quartiles), Freq)) + geom_bar(aes(group = cell.type, fill = cell.type), stat="identity", position = "stack") + theme_classic(base_size = 20) + theme(axis.line=element_line()) + facet_wrap(~ age.quartiles) + theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

 print(p4)
}

```

## Calculate cell type distribution score for each sample 

### Matrix, Table Implementaion
- Run CTDS function with matrix or table of normalized cell type proportions per sample and calculate the cell type distribution score per sample.
```{r}

div.res <- CTDS.score(as.matrix(ct_prop), metadata = metadat)

head(div.res)
```


### SingleCellExperiment Object, Seurat Object Implementation
- Run CTDS function with Seurat/SingleCellExperiment Object to calulate normalized cell type proportions per sample and calulate the cell type distribution score per sample.
```{r}
div.res <- CTDS.score(pbmc.combined, sample = "sample.ID", cell.type = "ct.consensus", metadata = metadat)

head(div.res)
```


### Visualization and statistical tests of cell type distribution scores 

Visualizations and statistical tests to compare cell type distribution scores between four age groups across the human lifespan: Younger (20-39), Middle Age (40-59), Older Age (60-89), and Extreme Longevity (100-119).

```{r, fig.dim=c(20,10)}
require(ggplot2)
plot.ctds <- ggplot(div.res, aes(age.quartiles, score)) + geom_boxplot(aes(fill = age.quartiles)) + theme_classic(base_size = 20) + theme(axis.line=element_line()) + ylim(0.8, 2.5)

plot.ctds

```

Statistical tests
- Anova and T-test of cell type distribution scores between the four age groups
```{r, fig.dim = c(20, 10)}
#Add control vs centenarian information
div.res$EL <- as.numeric(div.res$age.quartiles == "Extreme Longevity")

#histogram of cell type diversity values

#anova to test if there is a significant difference in at least one age group
lm.model <- lm(score ~ age.quartiles, data = div.res)
anova(lm.model)

#pairwsie t-tests for multiple comparisons between the scores of age groups
pairwise.t.test(div.res$score, div.res$age, p.adjust.method = "BH")

```

- Anova of cell type distribution scores between the three younger age groups
```{r}
#Filter out Extreme Longevity subjects
div.younger <- div.res %>% dplyr::filter(age.quartiles != "Extreme Longevity")

#anova to test if there is a significant difference in at least one age group
lm.younger <- lm(score ~ age.quartiles, data = div.younger)
anova(lm.younger)
```